File: r/diagonal_elements.R
library(sets)

diag_mult_nai <- function(A, B){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    bt <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- diag(ct)
    cache_scrub()
    start.time <- Sys.time()
    ct <- diag(at %*% bt)
    end.time <- Sys.time()

    return(tuple(as.double(end.time-start.time), ct))
}

diag_nai <- function(A, B){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    bt <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- diag(ct)
    cache_scrub()
 start.time <- Sys.time()
    ct <- diag(at + bt)
    end.time <- Sys.time()

    return(tuple(as.double(end.time-start.time), ct))
}

diag_rec <- function(A, B){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    bt <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- diag(ct)
    cache_scrub()
 start.time <- Sys.time()
    ct <- diag(at) + diag(bt)
    end.time <- Sys.time()

    return(tuple(as.double(end.time-start.time), ct))
}

diagonal_elements <- function(b, ...){

    res3 <- benchmark('diag_mult_nai', diag_mult_nai, ...)
    res1 <- benchmark('diag_add_nai', diag_nai, ...)
    res2 <- benchmark('diag_add_rec', diag_rec, ...)
    cat(c("DiagonalElements correctness: ", all.equal(res1, res2), "\n"), sep="")
}

File: python/diagonal_elements.py
from scipy import linalg
import numpy as np
import logging
import time
import copy
from benchmarker import cache_scrub

logger = logging.getLogger('DiagonalElements')

def diagonal_mult_nai(A, B):

    at = copy.deepcopy(A)
    bt = copy.deepcopy(B)
    ct = copy.deepcopy(B)
    ct = ct.diagonal()
    cache_scrub()
    start = time.perf_counter()
    ct = np.diagonal(at @ bt)
    end = time.perf_counter()

    return end-start, ct

def diagonal_nai(A, B):

    at = copy.deepcopy(A)
    bt = copy.deepcopy(B)
    ct = copy.deepcopy(B)
    ct = ct.diagonal()
    cache_scrub()
    start = time.perf_counter()
    ct = np.diagonal(at + bt)
    end = time.perf_counter()

    return end-start, ct


def diagonal_rec(A, B):

    at = copy.deepcopy(A)
    bt = copy.deepcopy(B)
    ct = copy.deepcopy(B)
    ct = ct.diagonal()
    cache_scrub()
    start = time.perf_counter()
    ct = np.diagonal(at) + np.diagonal(bt)
    end = time.perf_counter()

    return end-start, ct


def diagonal_elements(b, *args):

    res3 = b.benchmark('diag_mult_nai', diagonal_mult_nai, *args)
    res1 = b.benchmark('diag_add_nai', diagonal_nai, *args)
    res2 = b.benchmark('diag_add_rec', diagonal_rec, *args)
    logger.info('DiagonalElements correctness: {}'.format(np.allclose(res1, res2)))
